Datetime component

JavierCenteno opened the pull request.

JavierCenteno committed:
Add regexp service

JavierCenteno committed:
Merge branch 'devel' into i146-datetime-component

JavierCenteno committed:
Fix hasValue in input

JavierCenteno committed:
Implement datetime component
Fix #146

JavierCenteno committed:
Remove redundant setXInControl

JavierCenteno committed:
Remove icons

JavierCenteno committed:
Add includeSeconds property

JavierCenteno committed:
Document remaining properties of lux-datetime

JavierCenteno committed:
Set initial value of date

JavierCenteno committed:
Add example of includeSeconds property

JavierCenteno committed:
Add missing amperstand

pjmolina reviewed (changes requested):
1. Devuelve el bind `value` como un string formato Z:  
En lugar de: `2001-01-01T11:05:01`  11:05 en España debe ir relativo a Z:
`2001-01-01T09:05:01Z` y con informacion de zona horaria. Formato ISO8601 completo.
Sino, es ambiguo.

2. Veo cortado un poco la hora en chrome. No se si tiene arreglo.
3. Documetación: `Minimum date allowed. --> Minimum datetime allowed`  Idem para maximum

JavierCenteno committed:
Improve explanation of properties

JavierCenteno committed:
Don't remove timezone offset when stringifying dates

JavierCenteno committed:
Take timezone into account in datetime

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.scss:
```diff
@@ -0,0 +1,77 @@
+  padding: $input-padding;
+}
+
+.alert {
```
CS Refactor: move the .alert definition to styles/lux.scss  (the .alert is defined three times in lux). DRY
Look for common/repeated CSS and define only once.

pjmolina replied:
Same for .symbol, prefix, postfix ->define only once if posible and import style/lux

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.scss:
```diff
@@ -0,0 +1,77 @@
+
+.alert {
+  padding: 0.2rem 2rem;
+  background: #fe2e2e;
```
Main colors should be defined only once. Use a SASS variable in styles/lux.scss

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.ts:
```diff
@@ -0,0 +1,259 @@
+  @Output() valueChange = new EventEmitter<any>();
+  @Output() keyPress = new EventEmitter<KeyboardEvent>();
+
+  onChange = (value): void => {};
```
lint: rename to `_value`

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.ts:
```diff
@@ -0,0 +1,259 @@
+  }
+  // End of ControlValueAccessor Interface implementation
+
+  private setDateInControl(value: any): void {
```
Not used? remove

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.ts:
```diff
@@ -0,0 +1,259 @@
+  private setDateInControl(value: any): void {
+    this.dateInput.nativeElement.value = value;
+  }
+  private setTimeInControl(value: any): void {
```
Not used? remove

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.html:
```diff
@@ -0,0 +1,47 @@
+  [placeholder]=""
+
+  [ngClass]="className"
+  (focus)="markAsTouched()"
```
¿Porque marcamos como tocado al tomar el foco?
En lux.input solo lo hacemos al cambiar su valor o blur()
Seguir el mismo criterio en datetime y geoposition.

pjmolina replied:
En edición, cuando el valor no está completo debe devolver `null` como value y no `T`. Si devuelve un string -> debe ser un ISO8601 valido.

pjmolina replied:
El valor enlazado debe llevar formato ISO8601 con Z (zona horaria relativa a zona zulu) No la esta poniendo. (funcion `toIsoString()`  te lo da (no hace falta tomar los trozos)

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.html:
```diff
@@ -0,0 +1,47 @@
+  [step]="includeSeconds ? 1 : 60"
+
+  [ngClass]="className"
+  (focus)="markAsTouched()"
```
idem que el anterior.

pjmolina commented on projects/lux/src/lib/helperFns.ts:
```diff
@@ -25,6 +25,31 @@ export const normalizeDate = (value: any): string => {
+  new Date(date.getTime() - date.getTimezoneOffset() * 60000);
+
+export const dateToString = (date: Date): string => {
+  if (isNaN(date.getTime())) {
```
null protect? what `if (!date)`

pjmolina commented on projects/lux/src/lib/helperFns.ts:
```diff
@@ -25,6 +25,31 @@ export const normalizeDate = (value: any): string => {
+};
+
+export const dateToStringWithOffset = (date: Date): string => {
+  if (isNaN(date.getTime())) {
```
null check on date

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.ts:
```diff
@@ -0,0 +1,259 @@
+  @Input()
+  set value(v: string) {
+    const datetime = new Date(v);
+    const datetimeString = dateToString(datetime) + 'Z'; // YYYY-MM-DDThh:mm:ssZ
```
no uses un dateToString() custom - >usa `toIsoString()`

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.ts:
```diff
@@ -0,0 +1,259 @@
+
+  @Input()
+  set value(v: string) {
+    const datetime = new Date(v);
```
v null protect -> if null || undefined -> set null

pjmolina commented on projects/lux/src/lib/datetime/datetime.component.ts:
```diff
@@ -0,0 +1,259 @@
+    const initialAndEmpty = isInitialAndEmpty(this._value, v);
+    if (!isNaN(datetime.getTime())) {
+      this._value = datetimeString;
+      const offsetDatetimeString = dateToStringWithOffset(datetime); // YYYY-MM-DDThh:mm:ss
```
usa toIsoString()

pjmolina reviewed (changes requested):
Some improvements requested.

JavierCenteno committed:
Generalize common styles

JavierCenteno committed:
Fix style variables

JavierCenteno committed:
Import lux styles

JavierCenteno committed:
Generalize colors

JavierCenteno committed:
Remove unused methods

JavierCenteno committed:
Variable renaming

JavierCenteno committed:
Null checks

JavierCenteno committed:
Generalize isValidDate check

JavierCenteno committed:
Fix typo

JavierCenteno committed:
Use isValidDate function

JavierCenteno committed:
Remove dateToString

JavierCenteno committed:
Handle loss of focus

JavierCenteno committed:
Fix datetime binding

JavierCenteno committed:
improve null check

JavierCenteno committed:
Simplify function

JavierCenteno committed:
Set to undefined if provided values are empty

JavierCenteno committed:
Fix styles

JavierCenteno committed:
Clean up styles

pjmolina committed:
better defaults

pjmolina reviewed (changes requested):
Revisa el tratamiento de null.
Coge un valor y elimina componente a componente la fecha y la hora.
El valor enlazado debería ser **null**.

JavierCenteno committed:
Remove width from style

JavierCenteno committed:
Set value for includeSeconds example to 0

JavierCenteno committed:
Put includeSeconds before value so that seconds are not included when setting the value

JavierCenteno committed:
Add pipe to display null and undefined bind values in sample

JavierCenteno committed:
And sandbox form to datetime

pjmolina commented on src/app/datetime-sample/datetime-sample.component.ts:
```diff
@@ -16,7 +16,18 @@ export class DatetimeSampleComponent implements AfterContentInit {
   value3: string = new Date().toISOString();
   value4: string = new Date().toISOString();
   value5: string = new Date().toISOString().slice(0, 17) + '00.000Z';
```
Esto de `value5` es trampa :-) es preparar el valor en el cliente a sabiendas.

La prueba buena es **inicializar un valor con segundos distinto de cero -> y ver como el control (en el set) lo detecta y ajusta el valor a segundos == 0.**

La precisión que usemos, determina que se pierdan los segundos.

JavierCenteno committed:
Prune seconds if includeSeconds is false

JavierCenteno committed:
Use null for void values

JavierCenteno committed:
Handle null or undefined depending on case

JavierCenteno committed:
Prune miliseconds out of sample values

JavierCenteno committed:
Add clear button

JavierCenteno committed:
Add localTime variable to toggle whether we use local time or UTC

JavierCenteno committed:
Add button to toggle between UTC and local time

JavierCenteno committed:
Use a better minimum

JavierCenteno committed:
Make localTime a property

pjmolina committed:
css refactor. DRY
remove symbor border: use input border
ensure same size of border
fix infix postfix rounded
remove local time (only show UTC whe applicable) minimalism
add aria-label for clear

pjmolina commented:
Looks great to me. Merge and squashed to a single commit.
